docs/running-clusterfuzzlite/gitlab.md

```diff
@@ -0,0 +1,229 @@
+---
+layout: default
+parent: Step 2&#58; Running ClusterFuzzLite
+grand_parent: ClusterFuzzLite
+title: GitLab
+nav_order: 1
+permalink: /running-clusterfuzzlite/gitlab/
+---
+# GitLab
+{: .no_toc}
+
+- TOC
+{:toc}
+---
+
+This page explains how to set up ClusterFuzzLite to run on [GitLab].
+To get the most of this page, you should have already set up your
+[build integration] and read the more
+[high-level document on running ClusterFuzzLite].
+
+## Gitlab runner
+The following examples use a `docker` gitlab runner running sibling containers:
+[doc](https://docs.gitlab.com/ee/ci/docker/using_docker_build.html#use-docker-socket-binding).
+
+It should be possible to achieve the same functionalities with using a shell executor.
+But then, the `.gitlab-ci.yml` should be different, and explicitly call the `docker` commands
+on clusterfuzz-lite images.
+
+Docker-in-Docker does not seem possible as clusterfuzz-lite images
+do not have the docker tools installed.
+
+## .gitlab-ci.yml
+For basic ClusterFuzzLite functionality, all you need is a single job
+to enable fuzzing on your merge requests.
+
+To enable more features, we recommend having different jobs for:
+
+- continuous builds
+- batch fuzzing and corpus pruning
+- coverage
+
+### MR fuzzing
+
+To add a fuzzing job that fuzzes all merge requests to your repo, add the
+following default configurations to `.gitlab-ci.yml`:
+
+{% raw %}
+```yaml
+variables:
+  SANITIZER: "address"
+  WORKSPACE: $CI_PROJECT_DIR
+  REPOSITORY: $CI_PROJECT_NAME
+  CFL_PLATFORM: gitlab
+
+clusterfuzzlite:
+  image:
+    name: google/clusterfuzzlite/actions/build_fuzzers@v1
+    entrypoint: [""]
+  stage: fuzz
+  rules:
+    # default code change
+    - if: $CI_PIPELINE_SOURCE == "merge_request_event"
+      variables:
+        MODE: "code-change"
+  # to select a relevant runner
+  tags:
+    - fuzz
+  before_script:
+    # gitlab's container name is not its hostname !
+    - export CFL_CONTAINER_ID=`cut -c9- < /proc/1/cpuset`
+  script:
+    # will build and run the fuzzers
+    - python3 "/opt/oss-fuzz/infra/cifuzz/cifuzz_combined_entrypoint.py"
```

Comment:
Leaving this to Johnathan

Reply:
The images are exactly the same (for now) except for their entrypoints.
I'm not sure this is something I want to support and I don't like the fact that this is using build_fuzzers for running fuzzers.
I can publish a cifuzz-combined image that uses the combined entrypoint to support both.
but let me explain why we have different images.
On github, having two images makes it easy to seperate build logs from run logs.
See how in [these logs](https://github.com/systemd/systemd/runs/4704498147?check_suite_focus=true) you can click to view the build step and run step seperately.
Now, the drawback of this approach is that I think it has made the interface more complex, instead of having to invoke Cifuzz once it must be invoked twice. See how in this example two steps must be specified: https://google.github.io/clusterfuzzlite/running-clusterfuzzlite/github-actions/#pr-fuzzing

In hindsight, I think it might be a small mistake to have seperated the two actions. So the question is, should we have separate actions on gitlab as well (which will make things consistent at least?)

@oliverchang WDYT of all this?

Reply:
I see your point now. And I agree that it looks nicer in the logs. In GitLab the log view is different. You would have to specify two jobs per sanitizer, effectively it would look like this (shows 8 jobs):

<img width="627" alt="image" src="https://user-images.githubusercontent.com/8866185/148656886-34f008c9-17eb-4e3a-a15c-a8e157ffe9bc.png">

A click on each job leads to the job log.

Reply:
Not sure. As I mentioned, I think 2 actions/images makes the interface for users more complicated, but I guess that interface is only used once while the seperation provides benefits that are experienced many times.

Reply:
Sorry for the late reply. I think we should try to be as consistent as we can and keep the two separate entrypoints for gitlab as well. Does this add significantly more complexity to the gitlab integration?

Reply:
+1

@catenacyber can you please do this?

Reply:
ok will do

Reply:
Looking into this, I think this adds significant complexity, compared to already integrated CI systems.

We need to share the fuzzers built by `clusterfuzzlite-build-fuzzers` with `clusterfuzzlite-run-fuzzers`
As these are 2 different docker images, we need 2 different GitLab jobs. So they do not share any workspace...
To share data between jobs (no matter if they belong to the same pipeline or not), we either need cache or artifacts.

I am trying the solution where `clusterfuzzlite-build-fuzzers` ends with `cp -r $CI_BUILDS_DIR/$CI_JOB_ID/build-out/ $CI_PROJECT_DIR/artifacts/build-out/$SANITIZER` and `clusterfuzzlite-run-fuzzers` begins with `mv $CI_PROJECT_DIR/artifacts/build-out/$SANITIZER $CI_BUILDS_DIR/$CI_JOB_ID/build-out`
Anyways, the `build-out`contents needs to be put in a specific directory under the `$CI_PROJECT_DIR ` aka `project_src_path`. And we need to differentiate between the different sanitizers builds 

This can work also with cache instead of artifacts.
Note that I need to use `$CI_BUILDS_DIR/$CI_JOB_ID` as gitlab does not know `$WORKSPACE`
Another solution would be to bring back the workspace in the project directory... But we would still need to differentiate between the sanitizers...